### ISSUE TITLE
XLA leads to core dump

### ISSUE THREAD

**Issue by hzhangxyz** [tensorflow/tensorflow#12683]
### System information

[output of tf_env_collect.sh](http://paste.ubuntu.com/25424565/)

#### Tensorflow

Tensorflow compiled from the source v1.3.0(9e76bf3)

with cuda, with xla, without mpi, without mkl

#### OS
CentOS 7

out put of `uname -a`:

Linux zhanghao 3.10.0-514.26.2.el7.x86_64 #1 SMP Tue Jul 4 15:04:05 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux

#### python
Python 2.7.13 |Intel Corporation| (default, Apr 27 2017, 15:33:46)

[GCC 4.8.2 20140120 (Red Hat 4.8.2-15)] on linux2

#### Bezel
Build label: 0.5.2
Build target: bazel-out/local-fastbuild/bin/src/main/java/com/google/devtools/build/lib/bazel/BazelServer_deploy.jar
Build time: Tue Jun 27 13:27:03 2017 (1498570023)
Build timestamp: 1498570023
Build timestamp as int: 1498570023

#### GPU
CUDA 8.0 cuDNN 6.0.21
GPU: GeForce GTX 950M

### Describe the problem
core dump when use xla with gpu,

BTW, if use cpu only, xla won't lead to core dump

### Source code
This is code to reproduce the bug
```
import tensorflow as tf
import sys
D = 2
A = tf.random_normal(shape=[D, D, 2], dtype=tf.float32,name="A")
B = tf.random_normal(shape=[D, D, 2], dtype=tf.float32, name="B")
E = tf.ones(shape=[D], dtype=tf.float32, name="EBA")
H = tf.reshape(tf.constant([[0.25,0,0,0],[0,-0.25,0.5,0],[0,0.5,-0.25,0],[0,0,0,0.25]],
                           dtype=tf.float32),[2,2,2,2],name="Hamiltonian")
EA = tf.multiply(A,tf.reshape(E,[D,1,1]))
AB = tf.tensordot(EA,B,[[1],[0]],name="AB")
S, U, V = tf.svd(tf.reshape(AB,[2*D,2*D]))
UU = tf.transpose(tf.multiply(tf.reshape(U[:,:D],[D,2,D]),tf.reshape(E,[D,1,1])),[0,2,1],name="nA")
data = UU / tf.reduce_max(UU)
config = tf.ConfigProto()
if len(sys.argv)>1:
    config.graph_options.optimizer_options.global_jit_level = tf.OptimizerOptions.ON_1
sess = tf.Session(config=config)
sess.run(tf.global_variables_initializer())
print sess.run(data)
```
save as MPS.py and run `python MPS.py` and get:
```
2017-08-29 20:51:54.856662: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:893] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2017-08-29 20:51:54.857232: I tensorflow/core/common_runtime/gpu/gpu_device.cc:955] Found device 0 with properties:
name: GeForce GTX 950M
major: 5 minor: 0 memoryClockRate (GHz) 1.124
pciBusID 0000:0a:00.0
Total memory: 3.95GiB
Free memory: 3.92GiB
2017-08-29 20:51:54.857292: I tensorflow/core/common_runtime/gpu/gpu_device.cc:976] DMA: 0
2017-08-29 20:51:54.857301: I tensorflow/core/common_runtime/gpu/gpu_device.cc:986] 0:   Y
2017-08-29 20:51:54.857312: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1045] Creating TensorFlow device (/gpu:0) -> (device: 0, name: GeForce GTX 950M, pci bus id: 0000:0a:00.0)
[[[ 0.64684284 -0.48666194]
  [-0.56722689 -0.19181968]]

 [[ 0.34154066  0.77098727]
  [ 1.         -0.0881796 ]]]
```
and then run `python MPS.py -`, and get:
```
2017-08-29 20:52:18.127327: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:893] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2017-08-29 20:52:18.127719: I tensorflow/core/common_runtime/gpu/gpu_device.cc:955] Found device 0 with properties:
name: GeForce GTX 950M
major: 5 minor: 0 memoryClockRate (GHz) 1.124
pciBusID 0000:0a:00.0
Total memory: 3.95GiB
Free memory: 3.92GiB
2017-08-29 20:52:18.127780: I tensorflow/core/common_runtime/gpu/gpu_device.cc:976] DMA: 0
2017-08-29 20:52:18.127789: I tensorflow/core/common_runtime/gpu/gpu_device.cc:986] 0:   Y
2017-08-29 20:52:18.127799: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1045] Creating TensorFlow device (/gpu:0) -> (device: 0, name: GeForce GTX 950M, pci bus id: 0000:0a:00.0)
2017-08-29 20:52:18.319392: I tensorflow/compiler/xla/service/platform_util.cc:58] platform CUDA present with 1 visible devices
2017-08-29 20:52:18.319435: I tensorflow/compiler/xla/service/platform_util.cc:58] platform Executor present with 1 visible devices
2017-08-29 20:52:18.319739: I tensorflow/compiler/xla/service/platform_util.cc:58] platform Host present with 8 visible devices
2017-08-29 20:52:18.320615: I tensorflow/compiler/xla/service/service.cc:187] XLA service 0x7f59008c0350 executing computations on platform CUDA. Devices:
2017-08-29 20:52:18.320640: I tensorflow/compiler/xla/service/service.cc:195]   StreamExecutor device (0): GeForce GTX 950M, Compute Capability 5.0
2017-08-29 20:52:18.474432: F tensorflow/compiler/xla/util.cc:183] Check failed: p1.size() == p2.size() (3 vs. 0)
[1]    14077 abort (core dumped)  python MPS.py -
```

**Comment by jart** [comment]
@learyg Could you take a look at this XLA assertion failure coredump?

**Comment by sdvillal** [comment]
For what it is worth, I think I'm observing the same problem with our models. The bug seems to be present in both tensorflow 1.2.1 and 1.3.0, no matter how they were compiled - I have tested with a couple of variants. CUDA 8, CuDNN 6 and 1080Ti.

If useful I can upload a frozen version of one of these models.

**Comment by ygoncharov** [comment]
Same here

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by learyg** [comment]
Sorry for the seriously long delay, will try to repro. It's also helpful if you use TF_XLA_FLAGS="--xla_dump_computations_to=/tmp/" as a prefix on your cmdline and post the protobuf binary that appears as a result along with the git revision (the XLA compilation can be replayed directly with https://github.com/tensorflow/tensorflow/blob/e210cb140a60a74d5e9ce3bf9ebedb21b4910f1c/tensorflow/compiler/xla/tools/replay_computation.cc ).

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
A member of the TensorFlow organization has replied after the stat:awaiting tensorflower label was applied.

**Comment by tatatodd** [comment]
@jlebar @sanjoy do one of you have time to look into this issue?

**Comment by bixia1** [comment]
I am working on this issue.

**Comment by tatatodd** [comment]
Cool, thanks @bixia1 !  I think you might need to join the TensorFlow github organization in order to assign this issue to you.

In the meantime I'll set "contributions welcome" to avoid our bot (tensorflowbutler) from nagging us.

**Comment by bixia1** [comment]
I am able to reproduce the problem and verify that the latest tensorflow source doesn't have this problem. It looks to me that the following single line change can fix the problem, would you please give this a try? My source file may not be exactly the same as yours, so please manually apply the change to routine ConstantFolding::SimplifyGraph.
--- google/src/files/178965261/depot/google3/third_party/tensorflow/core/grappler/optimizers/constant_folding.cc
+++ google/src/cloud/bixia/ws17/google3/third_party/tensorflow/core/grappler/optimizers/constant_folding.cc
@@ -1357,7 +1357,7 @@ Status ConstantFolding::SimplifyGraph(GraphDef* output,
     const bool is_any_div = IsAnyDiv(*node);
     // Simplify multiplication by ones or zeros, and addition/subtraction of
     // zeros.
-    if (is_aggressive && use_shape_info &&
+    if (safe_to_use_shapes &&
         (is_mul || is_matmul || is_add || is_sub || is_any_div) &&
         properties.HasInputProperties(node->name()) &&
         properties.HasOutputProperties(node->name())) {

**Comment by hzhangxyz** [comment]
well, lastest tensorflow doesn't have this problem for this code indeed. but the following code still reproduces the similar core dump.

```
import tensorflow as tf
import sys
D = 2
A = tf.random_normal(shape=[D, D, 2], dtype=tf.float32,name="A")
B = tf.random_normal(shape=[D, D, 2], dtype=tf.float32, name="B")
if len(sys.argv)%2:
    fun = tf.random_normal
else:
    fun = tf.ones
E = fun(shape=[D], dtype=tf.float32, name="EBA")
H = tf.reshape(tf.constant([[0.25,0,0,0],[0,-0.25,0.5,0],[0,0.5,-0.25,0],[0,0,0,0.25]],
                           dtype=tf.float32),[2,2,2,2],name="Hamiltonian")
EA = tf.multiply(A,tf.reshape(E,[D,1,1]))
AB = tf.tensordot(EA,B,[[1],[0]],name="AB")
S, U, V = tf.svd(tf.reshape(AB,[2*D,2*D]))
UU = tf.transpose(tf.multiply(tf.reshape(U[:,:D],[D,2,D]),tf.reshape(E,[D,1,1])),[0,2,1],name="nA")
data = UU / tf.reduce_max(UU)
config = tf.ConfigProto()
if len(sys.argv)>2:
    config.graph_options.optimizer_options.global_jit_level = tf.OptimizerOptions.ON_1
sess = tf.Session(config=config)
sess.run(tf.global_variables_initializer())
print sess.run(data)
```

the only difference is changing tf.ones to tf.random_normal.

all of `python2 main.py`, `python2 main.py -`, `python2 main.py - - -` run correctly
but `python2 main.py - -` still core dump

:)

**Comment by jlebar** [comment]
Thanks for the bug report, @hzhangxyz.

Can you please file a separate bug and cc me and bixia@?  I don't expect this is the same issue as before, even though it looks very similar from your end.

It would be helpful if you could include the new log output in the new bug.  The first one wasn't a proper core dump but was instead an assertion failure; those log messages are useful to us.

Thanks again.
